File: schemas/pack-7/validate-fixtures.py
from pathlib import Path
from datetime import date, datetime
import json
import re

root = Path(__file__).resolve().parent

def resolve_ref(schema, reference):
    if not reference.startswith("#/"):
        raise ValueError(f"Unsupported external reference: {reference}")
    value = schema
    for part in reference[2:].split("/"):
        value = value[part.replace("~1", "/").replace("~0", "~")]
    return value

def validate(schema, rule, value, path="$"):
    errors = []
    if "$ref" in rule:
        return validate(schema, resolve_ref(schema, rule["$ref"]), value, path)
    if "required" in rule and isinstance(value, dict):
        for name in rule["required"]:
            if name not in value:
                errors.append(f"{path}.{name}: required property missing")
    if "const" in rule and value != rule["const"]:
        errors.append(f"{path}: must equal {rule['const']!r}")
    if "enum" in rule and value not in rule["enum"]:
        errors.append(f"{path}: value is not in the permitted catalogue")
    expected_type = rule.get("type")
    if expected_type == "object":
        if not isinstance(value, dict):
            return errors + [f"{path}: must be an object"]
        if rule.get("additionalProperties") is False:
            allowed = set(rule.get("properties", {}))
            for name in value:
                if name not in allowed:
                    errors.append(f"{path}.{name}: additional property not allowed")
        for name, child in value.items():
            if name in rule.get("properties", {}):
                errors.extend(validate(schema, rule["properties"][name], child, f"{path}.{name}"))
    elif expected_type == "array":
        if not isinstance(value, list):
            return errors + [f"{path}: must be an array"]
        if len(value) < rule.get("minItems", 0):
            errors.append(f"{path}: contains too few items")
        if rule.get("uniqueItems") and len({json.dumps(item, sort_keys=True) for item in value}) != len(value):
            errors.append(f"{path}: items must be unique")
        for index, child in enumerate(value):
            errors.extend(validate(schema, rule.get("items", {}), child, f"{path}[{index}]"))
    elif expected_type == "string":
        if not isinstance(value, str):
            return errors + [f"{path}: must be a string"]
        if len(value) < rule.get("minLength", 0):
            errors.append(f"{path}: string is too short")
        if len(value) > rule.get("maxLength", len(value)):
            errors.append(f"{path}: string is too long")
        if "pattern" in rule and not re.search(rule["pattern"], value):
            errors.append(f"{path}: string does not match required pattern")
        try:
            if rule.get("format") == "date-time":
                datetime.fromisoformat(value.replace("Z", "+00:00"))
            elif rule.get("format") == "date":
                date.fromisoformat(value)
        except ValueError:
            errors.append(f"{path}: invalid ISO 8601 {rule.get('format')}")
    elif expected_type == "integer":
        if not isinstance(value, int) or isinstance(value, bool):
            return errors + [f"{path}: must be an integer"]
        if value < rule.get("minimum", value):
            errors.append(f"{path}: below minimum")
    elif expected_type == "boolean":
        if not isinstance(value, bool):
            errors.append(f"{path}: must be a boolean")
    for branch_set in ("anyOf", "oneOf"):
        if branch_set in rule:
            passes = sum(not validate(schema, branch, value, path) for branch in rule[branch_set])
            if (branch_set == "anyOf" and passes < 1) or (branch_set == "oneOf" and passes != 1):
                errors.append(f"{path}: does not satisfy {branch_set}")
    return errors

results = []
failures = 0
for schema_path in sorted(root.glob("*.schema.json")):
    schema = json.loads(schema_path.read_text(encoding="utf-8"))
    if schema.get("$schema") != "https://json-schema.org/draft/2020-12/schema":
        raise ValueError(f"{schema_path.name} does not declare Draft 2020-12")
    stem = schema_path.name.removesuffix(".schema.json")
    for expected, folder in (("pass", "valid"), ("fail", "invalid")):
        fixture_path = root / "fixtures" / folder / f"{folder}-{stem}.json"
        instance = json.loads(fixture_path.read_text(encoding="utf-8"))
        errors = validate(schema, schema, instance)
        actual = "fail" if errors else "pass"
        passed = actual == expected
        failures += 0 if passed else 1
        results.append({"schema": schema_path.name, "fixture": fixture_path.name, "expected": expected, "actual": actual, "passed": passed, "errors": errors})

report = {"validator": "local deterministic validator for the Draft 2020-12 keywords used by this pack", "schemaSyntaxValid": True, "failures": failures, "results": results}
(root / "schema-validation-report.json").write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")
print(json.dumps(report, indent=2))
raise SystemExit(1 if failures else 0)

File: .codex-staging/pack-4-processing/04 Pack 4/Pack 4 Approved Export and Schema Draft/scripts/validate-fixtures.js
const fs=require('fs'); const path=require('path');
function check(schema, obj, p=''){
  let errors=[];
  if(schema.type==='object'){
    if(obj===null || Array.isArray(obj) || typeof obj!=='object') return [`${p||'/'}: expected object`];
    for(const r of (schema.required||[])) if(!(r in obj)) errors.push(`${p}/${r}: required property missing`);
    if(schema.additionalProperties===false){ for(const k of Object.keys(obj)) if(!schema.properties || !(k in schema.properties)) errors.push(`${p}/${k}: additional property not allowed`); }
    for(const [k,v] of Object.entries(obj)) if(schema.properties && schema.properties[k]) errors=errors.concat(check(schema.properties[k], v, `${p}/${k}`));
  } else if(schema.type==='array'){
    if(!Array.isArray(obj)) errors.push(`${p}: expected array`); else { if(schema.minItems && obj.length<schema.minItems) errors.push(`${p}: too few items`); obj.forEach((x,i)=>{ if(schema.items) errors=errors.concat(check(schema.items,x,`${p}/${i}`)); }); }
  } else if(schema.type==='string'){
    if(typeof obj!=='string') errors.push(`${p}: expected string`); if(schema.const!==undefined && obj!==schema.const) errors.push(`${p}: expected const ${schema.const}`); if(schema.enum && !schema.enum.includes(obj)) errors.push(`${p}: not in enum`); if(schema.minLength && obj.length<schema.minLength) errors.push(`${p}: too short`);
  } else if(schema.type==='integer'){
    if(!Number.isInteger(obj)) errors.push(`${p}: expected integer`); if(schema.minimum!==undefined && obj<schema.minimum) errors.push(`${p}: below minimum`);
  } else if(schema.type==='number'){
    if(typeof obj!=='number') errors.push(`${p}: expected number`); if(schema.exclusiveMinimum!==undefined && !(obj>schema.exclusiveMinimum)) errors.push(`${p}: not above exclusive minimum`);
  } else if(schema.type==='boolean'){
    if(typeof obj!=='boolean') errors.push(`${p}: expected boolean`); if(schema.const!==undefined && obj!==schema.const) errors.push(`${p}: expected const ${schema.const}`);
  }
  return errors;
}
const base=path.resolve(__dirname,'..','schemas'); let results=[]; let failures=0;
for(const sf of fs.readdirSync(base).filter(f=>f.endsWith('.schema.json')).sort()){
  const schema=JSON.parse(fs.readFileSync(path.join(base,sf),'utf8'));
  const stem=sf.replace('.schema.json','');
  for(const [kind, expected] of [['valid','pass'],['invalid','fail']]){
    const fp=path.join(base,'fixtures',kind,`${kind}-${stem}.json`);
    const obj=JSON.parse(fs.readFileSync(fp,'utf8'));
    const errors=check(schema,obj);
    const actual=errors.length?'fail':'pass';
    const passed=actual===expected; if(!passed) failures++;
    results.push({schema:sf, fixture:path.basename(fp), expected, actual, passed, errors});
  }
}
console.log(JSON.stringify({validator:'local structural subset validator', failures, results}, null, 2));
process.exit(failures?1:0);